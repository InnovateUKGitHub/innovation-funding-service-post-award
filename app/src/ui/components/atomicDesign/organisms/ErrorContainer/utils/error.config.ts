import React from "react";
import { UnauthenticatedError } from "../../../../../containers/pages/error/Unauthenticated/UnauthenticatedError";
import { GenericFallbackError } from "../../../../../containers/pages/error/Generic/GenericFallbackError";
import { NotFoundError } from "../../../../../containers/pages/error/NotFound/NotFoundError";

/**
 * @description Treat error messaging as unique key lookup, the type should be string.
 *
 * To ensure that we can scale the error handling, there maybe keys we do not have content for... In this we have a fallback
 */
export type InternalErrorTypes = "AUTHENTICATION_ERROR" | "NOT_FOUND";
export type ErrorTypes = InternalErrorTypes & "INTERNAL_ERROR_FALLBACK";

type ErrorConfig = Record<InternalErrorTypes, React.ElementType>;

export const internalErrorFallback: typeof GenericFallbackError = GenericFallbackError;

export const errorPages: ErrorConfig = {
  AUTHENTICATION_ERROR: UnauthenticatedError,
  NOT_FOUND: NotFoundError,
};
