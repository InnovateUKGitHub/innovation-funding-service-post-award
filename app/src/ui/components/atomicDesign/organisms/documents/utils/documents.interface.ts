import { DocumentSummaryDto } from "@framework/dtos/documentDto";

export interface DocumentsBase<T = DocumentSummaryDto> {
  documents: T[];
  qa: string;
}
