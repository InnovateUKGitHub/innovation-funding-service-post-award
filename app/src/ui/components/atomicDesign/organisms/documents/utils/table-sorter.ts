import { stringComparator, numberComparator, dateComparator } from "@framework/util/comparator";
import { Logger } from "@shared/developmentLogger";
import { useMounted } from "@ui/context/Mounted";
import { useMemo, useState } from "react";

export type SortOptions = "none" | "ascending" | "descending";

interface ColumnSortState<T> {
  sortKey?: T;
  sortDirection: SortOptions;
}

const logger = new Logger("Table Sort");

// Note: Get current state and return next state
const sortOptionsWorkflow: Record<SortOptions, SortOptions> = {
  none: "ascending",
  ascending: "descending",
  descending: "ascending",
};

/**
 * sorts rows according to the sort key and sort direction
 */
function sortRowsFromKey<T>({ sortKey, sortDirection }: ColumnSortState<keyof T | null | undefined>, rows: T[]) {
  if (!sortKey || sortDirection === "none") return rows;

  return rows.sort((unCheckedA, uncheckedB) => {
    const sortValues = [unCheckedA[sortKey], uncheckedB[sortKey]];

    if (sortDirection === "descending") sortValues.reverse();

    const [firstItem, secondItem] = sortValues;

    if (typeof firstItem === "string" && typeof secondItem === "string") {
      return stringComparator(firstItem, secondItem);
    }

    if (typeof firstItem === "number" && typeof secondItem === "number") {
      return numberComparator(firstItem, secondItem);
    }

    if (firstItem instanceof Date && secondItem instanceof Date) {
      return dateComparator(firstItem, secondItem);
    }

    logger.warn(`Table Sort value ${String(sortKey)} has not matched against any comparator`);
    return 0;
  });
}

/**
 * ### useTableSorter
 *
 * hook returns sorted table configuration
 */
export function useTableSorter<T>({
  sortKeys,
  tableRows,
  initialSortKey,
  initialSortState = "none",
}: {
  sortKeys: (keyof T | null | undefined)[];
  tableRows: T[];
  initialSortKey?: keyof T;
  initialSortState?: SortOptions;
}) {
  const { isServer } = useMounted();
  const [sortColumn, setSortColumn] = useState<ColumnSortState<keyof T | null | undefined>>({
    sortKey: initialSortKey,
    sortDirection: initialSortState,
  });

  const sortedRows = useMemo(() => sortRowsFromKey(sortColumn, tableRows), [sortColumn, tableRows]);

  const handleSort = (clickedIndex: number): void => {
    const sortKey = sortKeys[clickedIndex];

    if (isServer || !sortKey) return;

    setSortColumn(currentState => {
      const isDifferentSortKey = currentState.sortKey !== sortKey;
      const currentDirection = isDifferentSortKey ? initialSortState : sortColumn.sortDirection;
      const sortDirection = sortOptionsWorkflow[currentDirection];

      return {
        sortKey,
        sortDirection,
      };
    });
  };

  const getColumnOption = (columnIndex: number): SortOptions | undefined => {
    const sortKey = sortKeys[columnIndex];

    if (isServer || !sortColumn || !sortKey) return undefined;

    return sortColumn.sortKey === sortKey ? sortColumn.sortDirection : "none";
  };

  return {
    sortedRows,
    handleSort,
    getColumnOption,
  };
}
