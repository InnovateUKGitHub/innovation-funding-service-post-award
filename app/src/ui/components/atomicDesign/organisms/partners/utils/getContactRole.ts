import { PartnerDto } from "@framework/dtos/partnerDto";
import { ProjectContactDto } from "@framework/dtos/projectContactDto";

/**
 * Get the list of contact/partner pairs that match a specific partner role.
 *
 * @author Leondro Lio <leondro.lio@iuk.ukri.org>
 * @returns The list of contacts, matched to their corresponding partner for this specific partnerRole.
 */
export function getContactRole<
  C extends Pick<ProjectContactDto, "name" | "role" | "accountId">,
  P extends Pick<PartnerDto, "name" | "accountId">,
>({
  contacts,
  partners,
  partnerRole,
}: {
  contacts: C[];
  partners: P[];
  partnerRole: ProjectContactDto["role"];
}): { partner: P; contact: C }[] {
  return contacts
    .filter(contact => contact.role === partnerRole)
    .map(contact => {
      const partner = partners.find(partner => partner.accountId === contact.accountId);

      return {
        partner,
        contact,
      };
    })
    .filter(x => typeof x.partner !== undefined)
    .sort((a, b) => {
      const partnerNameA = a.partner?.name ?? "";
      const partnerNameB = b.partner?.name ?? "";

      const contactComparison = a.contact.name.localeCompare(b.contact.name);
      const partnerComparison = partnerNameA.localeCompare(partnerNameB);

      return partnerComparison || contactComparison;
    }) as { partner: P; contact: C }[];
}
