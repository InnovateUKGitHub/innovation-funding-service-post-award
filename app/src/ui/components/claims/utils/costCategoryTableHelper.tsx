import { CostCategoryType } from "@framework/constants/enums";
import { PCROrganisationType } from "@framework/constants/pcrConstants";
import { ClaimDto } from "@framework/dtos/claimDto";
import { CostCategoryDto } from "@framework/dtos/costCategoryDto";
import { CostsSummaryForPeriodDto } from "@framework/dtos/costsSummaryForPeriodDto";
import { PartnerDto } from "@framework/dtos/partnerDto";
import { ProjectDto } from "@framework/dtos/projectDto";
import { ILinkInfo } from "@framework/types/ILinkInfo";
import { roundCurrency, diffAsPercentage } from "@framework/util/numberHelper";
import { Link } from "@ui/components/links";
import { P } from "@ui/rhf-components/Typography";
import { Result } from "@ui/validation/result";

export interface ClaimProps {
  project: ProjectDto;
  partner: PartnerDto;
  costCategories: CostCategoryDto[];
  claim: ClaimDto;
  claimDetails: CostsSummaryForPeriodDto[];
  getLink: (costCategoryId: string) => ILinkInfo | null;
  validation?: Result;
}

export interface ClaimTableRow {
  isTotal: boolean;
  category: Pick<CostCategoryDto, "id" | "name">;
  cost: Pick<
    CostsSummaryForPeriodDto,
    | "costCategoryId"
    | "remainingOfferCosts"
    | "forecastThisPeriod"
    | "costsClaimedThisPeriod"
    | "offerTotal"
    | "costsClaimedToDate"
  >;
  label: string | JSX.Element;
  differenceInPounds: number;
  diffPercentage: number;
}

interface ClaimTableResponse {
  totalNegativeCategories: ClaimTableRow[];
  costCategories: ClaimTableRow[];
}

export interface ClaimTableProps {
  costCategories: Pick<CostCategoryDto, "id" | "name" | "competitionType" | "organisationType">[];
  project: Pick<ProjectDto, "competitionType">;
  partner: Pick<PartnerDto, "organisationType">;
  claimDetails: Pick<
    CostsSummaryForPeriodDto,
    | "costCategoryId"
    | "remainingOfferCosts"
    | "forecastThisPeriod"
    | "costsClaimedThisPeriod"
    | "offerTotal"
    | "costsClaimedToDate"
  >[];
  getLink: (costCategoryId: string) => ILinkInfo | null;
  validation?: Result;
  disabled?: boolean;
}
/**
 * creates the table data
 */
export function createTableData(props: ClaimTableProps): ClaimTableResponse {
  const costCategories: ClaimTableRow[] = [];
  const totalNegativeCategories: ClaimTableRow[] = [];

  const filteredCategories = props.costCategories.filter(x => {
    const isMatchingCompetition = x.competitionType === props.project.competitionType;
    const isMatchingOrg = x.organisationType === props.partner.organisationType;

    return isMatchingCompetition && isMatchingOrg;
  });

  for (const category of filteredCategories) {
    const row = createRow(category, props, props.disabled);

    costCategories.push(row.costCategory);

    if (row.hasNegativeCost) {
      totalNegativeCategories.push(row.costCategory);
    }
  }

  const totalRow = calculateTotalRow(props.claimDetails);
  const allCostCategories = [...costCategories, totalRow];

  return {
    totalNegativeCategories,
    costCategories: allCostCategories,
  };
}

const emptyCostsSummaryForPeriodDto: CostsSummaryForPeriodDto = {
  costCategoryId: "",
  offerTotal: 0,
  forecastThisPeriod: 0,
  costsClaimedToDate: 0,
  costsClaimedThisPeriod: 0,
  remainingOfferCosts: 0,
};

/**
 * calculates total for the row
 */
function calculateTotalRow(claimDetails: ClaimTableProps["claimDetails"]): ClaimTableRow {
  let totalRowCosts: CostsSummaryForPeriodDto = {
    costCategoryId: "",
    offerTotal: 0,
    forecastThisPeriod: 0,
    costsClaimedToDate: 0,
    costsClaimedThisPeriod: 0,
    remainingOfferCosts: 0,
  };

  // Note: Populate totals form all claim items
  for (const item of claimDetails) {
    totalRowCosts = {
      costCategoryId: "",
      offerTotal: totalRowCosts.offerTotal + item.offerTotal,
      forecastThisPeriod: totalRowCosts.forecastThisPeriod + item.forecastThisPeriod,
      costsClaimedToDate: totalRowCosts.costsClaimedToDate + item.costsClaimedToDate,
      costsClaimedThisPeriod: totalRowCosts.costsClaimedThisPeriod + item.costsClaimedThisPeriod,
      remainingOfferCosts: totalRowCosts.remainingOfferCosts + item.remainingOfferCosts,
    };
  }

  const staticTotalRowData = {
    isTotal: true,
    label: "Total",
    category: {
      name: "Total",
      type: CostCategoryType.Unknown,
      id: "",
      isCalculated: false,
      hasRelated: false,
      competitionType: "Unknown",
      organisationType: PCROrganisationType.Unknown,
      description: "",
      hintText: "",
    },
  };

  return {
    ...staticTotalRowData,
    cost: totalRowCosts,
    differenceInPounds: roundCurrency(totalRowCosts.forecastThisPeriod - totalRowCosts.costsClaimedThisPeriod),
    diffPercentage: diffAsPercentage(totalRowCosts.forecastThisPeriod, totalRowCosts.costsClaimedThisPeriod),
  };
}

/**
 * creates a row
 */
function createRow(
  category: Pick<CostCategoryDto, "id" | "name">,
  claimItem: {
    claimDetails: Pick<
      CostsSummaryForPeriodDto,
      | "costCategoryId"
      | "remainingOfferCosts"
      | "forecastThisPeriod"
      | "costsClaimedThisPeriod"
      | "offerTotal"
      | "costsClaimedToDate"
    >[];
    getLink: ClaimTableProps["getLink"];
    validation?: ClaimTableProps["validation"];
  },
  disabled?: boolean,
) {
  const { claimDetails, getLink, validation } = claimItem;

  // TODO: ID will always be present... 🤷🏻‍♂️
  // Note: This function may not have context of any claim details, so we provide a default
  const item = claimDetails.find(x => x.costCategoryId === category.id);

  const hasNegativeCost: boolean = !!item && item.remainingOfferCosts < 0;

  const costCategory = {
    label: renderCostCategory({ getLink, validation }, category, disabled),
    isTotal: false,
    category,
    cost: item || emptyCostsSummaryForPeriodDto,
    differenceInPounds: item ? roundCurrency(item.forecastThisPeriod - item.costsClaimedThisPeriod) : 0,
    diffPercentage: item ? diffAsPercentage(item.forecastThisPeriod, item.costsClaimedThisPeriod) : 0,
  };

  return {
    hasNegativeCost,
    costCategory,
  };
}

export type ClaimInfoProps = Pick<ClaimProps, "getLink" | "validation">;
export type CategoryInfoProps = Pick<CostCategoryDto, "id" | "name">;

export const renderCostCategory = (claimInfo: ClaimInfoProps, categoryInfo: CategoryInfoProps, disabled?: boolean) => {
  const { getLink, validation } = claimInfo;
  const route = getLink(categoryInfo.id);

  if (!route) return categoryInfo.name;

  const linkId = (validation && validation.errorMessage && validation.key) || "";
  if (disabled) {
    return <P id={linkId}>{categoryInfo.name}</P>;
  }
  return (
    <Link id={linkId} route={route}>
      {categoryInfo.name}
    </Link>
  );
};
