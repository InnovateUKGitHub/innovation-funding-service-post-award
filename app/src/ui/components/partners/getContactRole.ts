import { PartnerDto, ProjectContactDto } from "@framework/dtos";

export interface IContactRole {
  partner?: PartnerDto;
  contact: ProjectContactDto;
}

/**
 * Get the list of contact/partner pairs that match a specific partner role.
 *
 * @author Leondro Lio <leondro.lio@iuk.ukri.org>
 * @returns The list of contacts, matched to their corresponding partner for this specific partnerRole.
 */
export const getContactRole = ({
  contacts,
  partners,
  partnerRole,
}: {
  contacts: ProjectContactDto[];
  partners: PartnerDto[];
  partnerRole: ProjectContactDto["role"];
}) =>
  contacts
    .filter(contact => contact.role === partnerRole)
    .map(contact => {
      const partner = partners.find(partner => partner.accountId === contact.accountId);

      return {
        partner,
        contact,
      };
    })
    .sort((a, b) => {
      const partnerNameA = a.partner?.name ?? "";
      const partnerNameB = b.partner?.name ?? "";

      const contactComparison = a.contact.name.localeCompare(b.contact.name);
      const partnerComparison = partnerNameA.localeCompare(partnerNameB);

      return partnerComparison || contactComparison;
    });
