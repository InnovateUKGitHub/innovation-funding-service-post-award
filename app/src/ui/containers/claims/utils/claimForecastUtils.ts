import { ForecastDetailsDTO, ProjectDto } from "@framework/dtos";

// TODO: ACC-7461
export const getArrayFromPeriod = (
  originalArray: ForecastDetailsDTO[],
  currentPeriod: ProjectDto["periodId"],
  lastPeriodId: number,
): ForecastDetailsDTO[] => {
  if (originalArray.length === 0) return [];
  return originalArray.filter(
    forecast => forecast.periodId >= currentPeriod && forecast.periodId <= lastPeriodId,
  );
};
