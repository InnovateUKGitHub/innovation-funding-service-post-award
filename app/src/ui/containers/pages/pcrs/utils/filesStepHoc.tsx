import { ContentSelector } from "@copy/type";
import { DocumentDescription } from "@framework/constants/documentDescription";
import { DocumentSummaryDto } from "@framework/dtos/documentDto";
import { MultipleDocumentUploadDto } from "@framework/dtos/documentUploadDto";
import { PCRItemDto } from "@framework/dtos/pcrDtos";
import { Content } from "@ui/components/atomicDesign/molecules/Content/content";
import { Section } from "@ui/components/atomicDesign/molecules/Section/section";
import { DocumentGuidance } from "@ui/components/atomicDesign/organisms/documents/DocumentGuidance/DocumentGuidance";
import { DocumentEdit } from "@ui/components/atomicDesign/organisms/documents/DocumentView/DocumentView";
import { createTypedForm } from "@ui/components/bjss/form/form";
import { Loader } from "@ui/components/bjss/loading";
import { PcrStepProps } from "@ui/containers/pages/pcrs/pcrWorkflow";
import { useContent } from "@ui/hooks/content.hook";
import { EditorStatus } from "@ui/redux/constants/enums";
import { useStores } from "@ui/redux/storesProvider";

interface FilesStepProps {
  isSaving: boolean;
  documents: DocumentSummaryDto[];
  onFileChange: (
    {
      saving,
      continuing,
    }: {
      saving: boolean;
      continuing: boolean;
    },
    dto: MultipleDocumentUploadDto,
  ) => void;
  onFileDelete: (dto: MultipleDocumentUploadDto, document: DocumentSummaryDto) => void;
}

const WithPcrFilesStep = <TDto extends PCRItemDto, TValidator>({
  heading,
  guidance,
  documentDescription,
}: {
  heading: ContentSelector;
  guidance?: ContentSelector;
  documentDescription: DocumentDescription;
}) => {
  const Form = createTypedForm<TDto>();
  const UploadForm = createTypedForm<MultipleDocumentUploadDto>();

  const PcrFilesStep = (props: PcrStepProps<TDto, TValidator> & FilesStepProps) => {
    const { documents, documentsEditor, isSaving } = props;

    return (
      <>
        <Section>
          <UploadForm.Form
            enctype="multipart"
            editor={documentsEditor}
            onSubmit={() => props.onFileChange({ saving: true, continuing: true }, documentsEditor.data)}
            onChange={dto => props.onFileChange({ saving: false, continuing: false }, dto)}
            qa="projectChangeRequestItemUpload"
            disabled={isSaving}
          >
            <UploadForm.Fieldset heading={heading}>
              {guidance && <Content markdown value={guidance} />}

              <UploadForm.Hidden name="description" value={() => documentDescription} />

              <DocumentGuidance />

              <UploadForm.MultipleFileUpload
                label={x => x.documentLabels.uploadInputLabel}
                name="attachment"
                labelHidden
                value={data => data.files}
                update={(dto, files) => {
                  dto.files = files || [];
                  dto.description = documentDescription;
                }}
                validation={documentsEditor.validator.files}
              />
            </UploadForm.Fieldset>

            <UploadForm.Fieldset>
              <UploadForm.Button
                name="uploadFile"
                styling="Secondary"
                onClick={() => props.onFileChange({ saving: true, continuing: false }, documentsEditor.data)}
              >
                <Content value={x => x.pcrItem.uploadDocumentsButton} />
              </UploadForm.Button>
            </UploadForm.Fieldset>
          </UploadForm.Form>
        </Section>

        <Section>
          <DocumentEdit
            qa="prepare-item-file-for-partner-documents"
            onRemove={document => props.onFileDelete(documentsEditor.data, document)}
            documents={documents}
          />
        </Section>

        <Form.Form
          qa="saveAndContinue"
          data={props.pcrItem}
          // Reset any document errors by resetting files to "[]".
          onSubmit={() => props.onFileChange({ saving: false, continuing: true }, { files: [] })}
          disabled={isSaving}
        >
          <Form.Fieldset>
            <Form.Button name="default" styling="Primary">
              <Content value={x => x.pcrItem.submitButton} />
            </Form.Button>
          </Form.Fieldset>
        </Form.Form>
      </>
    );
  };

  const PcrFilesStepContainer = (props: PcrStepProps<TDto, TValidator>) => {
    const stores = useStores();
    const { getContent } = useContent();

    return (
      <Loader
        pending={stores.projectChangeRequestDocuments.pcrOrPcrItemDocuments(props.project.id, props.pcrItem.id)}
        render={documents => (
          <PcrFilesStep
            {...props}
            isSaving={props.status === EditorStatus.Saving || props.documentsEditor.status === EditorStatus.Saving}
            documents={documents}
            onFileChange={({ saving, continuing }, dto) => {
              stores.messages.clearMessages();
              // show message if remaining on page
              const successMessage = saving
                ? getContent(x => x.documentMessages.uploadDocuments({ count: dto.files.length }))
                : undefined;

              // Save the documents editor.
              stores.projectChangeRequestDocuments.updatePcrOrPcrItemDocumentsEditor(
                saving,
                props.project.id,
                props.pcrItem.id,
                dto,
                saving,
                successMessage,
                () => {
                  if (saving && continuing) {
                    props.onSave(false);
                  }
                },
              );

              if (!saving && continuing) {
                props.onSave(false);
              }
            }}
            onFileDelete={(dto, document) => {
              stores.messages.clearMessages();
              stores.projectChangeRequestDocuments.deletePcrOrPcrItemDocumentsEditor(
                props.project.id,
                props.pcrItem.id,
                dto,
                document,
                getContent(x => x.documentMessages.deletedDocument({ deletedFileName: document.fileName })),
              );
            }}
          />
        )}
      />
    );
  };

  return PcrFilesStepContainer;
};

export { WithPcrFilesStep };
