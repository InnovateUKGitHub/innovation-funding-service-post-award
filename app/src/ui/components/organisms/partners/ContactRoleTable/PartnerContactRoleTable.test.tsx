import { render } from "@testing-library/react";
import { TestBed } from "@shared/TestBed";

import {
  PartnerContactRoleTable,
  PartnersAndFinanceContactsProps,
} from "@ui/components/organisms/partners/ContactRoleTable/PartnerContactRoleTable";
import { initStubTestIntl } from "@shared/initStubTestIntl";
import { PartnerDto } from "@framework/dtos/partnerDto";
import { ProjectContactDto } from "@framework/dtos/projectContactDto";
import { getContactRole } from "../utils/getContactRole";

const testPartnerData: PartnerDto[] = [
  {
    id: "123",
    isLead: true,
    projectId: "456" as ProjectId,
    accountId: "321",
    name: "Steel Manufacturing",
  },
  {
    id: "912",
    isLead: false,
    projectId: "456" as ProjectId,
    accountId: "312",
    name: "University of Life",
  },
  {
    id: "431",
    isLead: false,
    projectId: "456" as ProjectId,
    accountId: "832",
    name: "Driverless Cars",
  },
] as PartnerDto[];

const testContactData: Omit<ProjectContactDto, "firstName" | "lastName">[] = [
  {
    id: "100" as ProjectContactLinkId,
    name: "Ted Tester",
    role: "Finance contact",
    roleName: "Finance Contact",
    email: "tedtester@nowhere.com",
    contactId: "908" as ContactId,
    accountId: "321" as AccountId,
    projectId: "456" as ProjectId,
    startDate: null,
    endDate: null,
    associateStartDate: null,
  },
  {
    id: "101" as ProjectContactLinkId,
    name: "Dave Developer",
    role: "Project Manager",
    roleName: "Project Manager",
    email: "davedeveloper@nowhere.com",
    contactId: "908" as ContactId,
    accountId: "312" as AccountId,
    projectId: "456" as ProjectId,
    startDate: null,
    endDate: null,
    associateStartDate: null,
  },
  {
    id: "102" as ProjectContactLinkId,
    name: "Joe Bloggs",
    role: "Finance contact",
    roleName: "Finance Contact",
    email: "joebloggs@nowhere.com",
    contactId: "908" as ContactId,
    accountId: "312" as AccountId,
    projectId: "456" as ProjectId,
    startDate: null,
    endDate: null,
    associateStartDate: null,
  },
  {
    id: "103" as ProjectContactLinkId,
    name: "Bob Baker",
    role: "Finance contact",
    roleName: "Finance Contact",
    email: "bobbaker@nowhere.com",
    contactId: "908" as ContactId,
    accountId: "832" as AccountId,
    projectId: "456" as ProjectId,
    startDate: null,
    endDate: null,
    associateStartDate: null,
  },
];

describe("<PartnersAndFinanceContacts />", () => {
  const setup = (partnerRole: ProjectContactDto["role"]) => {
    const defaultProps: PartnersAndFinanceContactsProps = {
      contactRoles: getContactRole({
        contacts: testContactData,
        partners: testPartnerData,
        partnerRole,
      }),
      qa: "partner-contact-role-table",
    };

    return render(
      <TestBed>
        <PartnerContactRoleTable {...defaultProps} />
      </TestBed>,
    );
  };

  beforeAll(async () => {
    await initStubTestIntl({
      projectContactLabels: {
        contactEmail: "Email",
        contactName: "Name",
        partnerName: "Partner",
      },
    });
  });

  describe("@renders", () => {
    test("matching MO snapshot", () => {
      expect(setup("Monitoring officer").container).toMatchSnapshot();
    });

    test("matching FC snapshot", () => {
      expect(setup("Finance contact").container).toMatchSnapshot();
    });

    test("matching PM snapshot", () => {
      expect(setup("Project Manager").container).toMatchSnapshot();
    });

    test("with partners names and Steel Manufacturing as the lead", () => {
      const { queryByText } = setup("Finance contact");

      const partnerNames = [
        queryByText("Steel Manufacturing (Lead)"),
        queryByText("University of Life"),
        queryByText("Driverless Cars"),
      ];

      partnerNames.forEach(partnerName => expect(partnerName).toBeInTheDocument());
    });

    test("with Finance Contacts", () => {
      const { queryByText } = setup("Finance contact");

      expect(queryByText("Ted Tester")).toBeInTheDocument();
      expect(queryByText("Joe Bloggs")).toBeInTheDocument();
      expect(queryByText("Bob Baker")).toBeInTheDocument();
    });

    test("with emails of all Finance Contact", () => {
      const { queryByText } = setup("Finance contact");

      expect(queryByText("tedtester@nowhere.com")).toBeInTheDocument();
      expect(queryByText("joebloggs@nowhere.com")).toBeInTheDocument();
      expect(queryByText("bobbaker@nowhere.com")).toBeInTheDocument();
    });

    test("with Project Managers", () => {
      const { queryByText } = setup("Project Manager");

      expect(queryByText("Dave Developer")).toBeInTheDocument();
    });

    test("with emails of all Project Managers", () => {
      const { queryByText } = setup("Project Manager");

      expect(queryByText("davedeveloper@nowhere.com")).toBeInTheDocument();
    });
  });
});
