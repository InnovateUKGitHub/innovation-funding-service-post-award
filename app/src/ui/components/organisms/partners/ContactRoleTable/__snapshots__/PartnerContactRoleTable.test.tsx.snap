// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PartnersAndFinanceContacts /> @renders matching FC snapshot 1`] = `
<div>
  <div
    class="govuk-table-wrapper"
    data-qa="partner-contact-role-table"
  >
    <table
      class="govuk-table"
    >
      <caption
        class="govuk-visually-hidden"
      >
        stub-caption
      </caption>
      <colgroup>
        <col
          data-qa="col-fc-name"
        />
        <col
          data-qa="col-partner-name"
        />
        <col
          data-qa="col-fc-email"
        />
      </colgroup>
      <thead
        class="govuk-table__head"
      >
        <tr
          class="govuk-table__row"
        >
          <th
            class="govuk-table__header"
            scope="col"
          >
            Name
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Partner
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Email
          </th>
        </tr>
      </thead>
      <tbody
        class="govuk-table__body"
      >
        <tr
          class="govuk-table__row"
        >
          <td
            class="govuk-table__cell"
          >
            Bob Baker
          </td>
          <td
            class="govuk-table__cell"
          >
            Driverless Cars
          </td>
          <td
            class="govuk-table__cell"
          >
            <a
              class="govuk-link govuk-!-font-size-19"
              href="mailto:bobbaker@nowhere.com"
            >
              bobbaker
              <wbr />
              @
              nowhere.com
            </a>
          </td>
        </tr>
        <tr
          class="govuk-table__row"
        >
          <td
            class="govuk-table__cell"
          >
            Ted Tester
          </td>
          <td
            class="govuk-table__cell"
          >
            Steel Manufacturing (Lead)
          </td>
          <td
            class="govuk-table__cell"
          >
            <a
              class="govuk-link govuk-!-font-size-19"
              href="mailto:tedtester@nowhere.com"
            >
              tedtester
              <wbr />
              @
              nowhere.com
            </a>
          </td>
        </tr>
        <tr
          class="govuk-table__row"
        >
          <td
            class="govuk-table__cell"
          >
            Joe Bloggs
          </td>
          <td
            class="govuk-table__cell"
          >
            University of Life
          </td>
          <td
            class="govuk-table__cell"
          >
            <a
              class="govuk-link govuk-!-font-size-19"
              href="mailto:joebloggs@nowhere.com"
            >
              joebloggs
              <wbr />
              @
              nowhere.com
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<PartnersAndFinanceContacts /> @renders matching MO snapshot 1`] = `
<div>
  <div
    class="govuk-table-wrapper"
    data-qa="partner-contact-role-table"
  >
    <table
      class="govuk-table"
    >
      <caption
        class="govuk-visually-hidden"
      >
        stub-caption
      </caption>
      <colgroup>
        <col
          data-qa="col-fc-name"
        />
        <col
          data-qa="col-partner-name"
        />
        <col
          data-qa="col-fc-email"
        />
      </colgroup>
      <thead
        class="govuk-table__head"
      >
        <tr
          class="govuk-table__row"
        >
          <th
            class="govuk-table__header"
            scope="col"
          >
            Name
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Partner
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Email
          </th>
        </tr>
      </thead>
      <tbody
        class="govuk-table__body"
      />
    </table>
  </div>
</div>
`;

exports[`<PartnersAndFinanceContacts /> @renders matching PM snapshot 1`] = `
<div>
  <div
    class="govuk-table-wrapper"
    data-qa="partner-contact-role-table"
  >
    <table
      class="govuk-table"
    >
      <caption
        class="govuk-visually-hidden"
      >
        stub-caption
      </caption>
      <colgroup>
        <col
          data-qa="col-fc-name"
        />
        <col
          data-qa="col-partner-name"
        />
        <col
          data-qa="col-fc-email"
        />
      </colgroup>
      <thead
        class="govuk-table__head"
      >
        <tr
          class="govuk-table__row"
        >
          <th
            class="govuk-table__header"
            scope="col"
          >
            Name
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Partner
          </th>
          <th
            class="govuk-table__header"
            scope="col"
          >
            Email
          </th>
        </tr>
      </thead>
      <tbody
        class="govuk-table__body"
      >
        <tr
          class="govuk-table__row"
        >
          <td
            class="govuk-table__cell"
          >
            Dave Developer
          </td>
          <td
            class="govuk-table__cell"
          >
            University of Life
          </td>
          <td
            class="govuk-table__cell"
          >
            <a
              class="govuk-link govuk-!-font-size-19"
              href="mailto:davedeveloper@nowhere.com"
            >
              davedeveloper
              <wbr />
              @
              nowhere.com
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
